pkg/pool-hooks/contracts/StableSurgeHook.sol

```diff
@@ -0,0 +1,338 @@
+    }
+
+    /// @dev Ensure the sender is the swapFeeManager, or default to governance if there is no manager.
+    function _ensureValidSender(address pool) private view {
+        address swapFeeManager = _vault.getPoolRoleAccounts(pool).swapFeeManager;
+
+        if (swapFeeManager == address(0)) {
+            if (_canPerform(getActionId(msg.sig), msg.sender, pool) == false) {
+                revert SenderNotAllowed();
+            }
+        } else if (swapFeeManager != msg.sender) {
+            revert SenderNotAllowed();
+        }
+    }
+
+    /**
+     * @notice Calculate the surge fee percentage. If below threshold, return the standard static swap fee percentage.
+     * @dev It is public to allow it to be called off-chain.
+     * @param params Input parameters for the swap (balances needed)
+     * @param pool The pool we are computing the fee for
+     * @param staticFeePercentage The static fee percentage for the pool (default if there is no surge)
+     */
+    function getSurgeFeePercentage(
+        PoolSwapParams calldata params,
+        address pool,
+        uint256 staticFeePercentage
+    ) public view returns (uint256) {
+        uint256 numTokens = params.balancesScaled18.length;
+
+        uint256 invariant = IBasePool(pool).computeInvariant(params.balancesScaled18, Rounding.ROUND_DOWN);
+        (uint256 currentAmp, , ) = IStablePool(pool).getAmplificationParameter();
+
+        uint256 amountCalculatedScaled18;
+        if (params.kind == SwapKind.EXACT_IN) {
+            amountCalculatedScaled18 = StableMath.computeOutGivenExactIn(
+                currentAmp,
+                params.balancesScaled18,
+                params.indexIn,
+                params.indexOut,
+                params.amountGivenScaled18,
+                invariant
+            );
+        } else {
+            amountCalculatedScaled18 = StableMath.computeInGivenExactOut(
+                currentAmp,
+                params.balancesScaled18,
+                params.indexIn,
+                params.indexOut,
+                params.amountGivenScaled18,
+                invariant
+            );
+        }
+
+        uint256[] memory newBalances = new uint256[](numTokens);
+        ScalingHelpers.copyToArray(params.balancesScaled18, newBalances);
+
+        for (uint256 i = 0; i < numTokens; ++i) {
+            if (params.kind == SwapKind.EXACT_IN) {
+                if (i == params.indexIn) {
+                    newBalances[i] -= params.amountGivenScaled18;
+                } else if (i == params.indexOut) {
+                    newBalances[i] += amountCalculatedScaled18;
+                }
+            } else {
+                if (i == params.indexIn) {
+                    newBalances[i] += amountCalculatedScaled18;
+                } else if (i == params.indexOut) {
+                    newBalances[i] -= params.amountGivenScaled18;
+                }
+            }
+        }
```

Comment:
I think we don't need the loop.
Also I think the signs for exact in are upside down. We always need to add for index in, and always need to subtract for index out; it's a matter of which one we use in each case.

```suggestion
        if (params.kind == SwapKind.EXACT_IN) {
            newBalances[params.indexIn] += params.amountGivenScaled18;
            newBalances[params.indexOut] -= amountCalculatedScaled18;
        } else {
            newBalances[params.indexIn] += amountCalculatedScaled18;
            newBalances[params.indexOut] -= params.amountGivenScaled18;
        }
```